### PR TITLE
Implement BatchNormGradient kernel for CPU EP

### PR DESCRIPTION
**Pull request by pranav-prakash**
**Description**: Register an implementation for BatchNormInternal and add a CPU kernel for BatchNormGradient. This is the third in a series of PRs to implement BN training on CPU (first was #6946, second was #7539).

**Motivation and Context**
Support training networks with BatchNorm (e.g. convnets). Also note that there exists a CUDA kernel for BN (forward training & backwards) but it's currently disabled due to flaky failures; someone more familiar with those parts can register the implementation for BNInternal on CUDA (gradient kernel doesn't have to change).

### CODE REVIEW

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+namespace onnxruntime {
+namespace contrib {
+
+template <typename T>
+Status BatchNormalizationGrad<T>::Compute(OpKernelContext* ctx) const {
+  const Tensor* dY = ctx->Input<Tensor>(0);
+  const Tensor* X = ctx->Input<Tensor>(1);
+  const Tensor* Scale = ctx->Input<Tensor>(2);
+  const Tensor* saved_mean = ctx->Input<Tensor>(3);
+  const Tensor* saved_inv_variance = ctx->Input<Tensor>(4);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:35]: Just noticed a minor thing on BatchNormInternal's schema, 
.Output(4, "saved_inv_std", "Inverse standard deviation for the batch", "U", OpSchema::Optional, true, 1, OpSchema::NonDifferentiable) 


Output4's name should actually be save_inv_variance.

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+  const auto* X_data = X->template Data<T>();
+  const auto* Scale_data = Scale->template Data<T>();
+  const auto* saved_mean_data = saved_mean->template Data<T>();
+  const auto* saved_inv_variance_data = saved_inv_variance->template Data<T>();
+
+  auto* dX_data = ctx->Output(0, input_shape)->template MutableData<T>();
+  auto* dScale_data = ctx->Output(1, channel_shape)->template MutableData<T>();
+  auto* dBias_data = ctx->Output(2, channel_shape)->template MutableData<T>();
+
+  const TensorShape& x_shape = X->Shape();
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:53]: duplicate of input_shape?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  const auto& dims_vec = x_shape.GetDims();
+  const size_t N = dims_vec[0];
+  const size_t C = dims_vec[1];  // assume NCHW as per the spec
+
+  // calculate sample_size (per individual channel)
+  size_t sample_size = 1;
+  for (size_t i = 2; i < dims_vec.size(); ++i) {
+    sample_size *= gsl::narrow<size_t>(dims_vec[i]);
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:61]: can use x_shape.SizeFromDimension(2)

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  for (size_t nc = 0; nc < N * C; ++nc) {
+    int c = nc % C;
+    dBias_arr(c) += dY_arr.col(nc).sum();
+    dScale_arr(c) +=
+        ((X_arr.col(nc) - mean_arr(c)) * inv_var_arr(c) * dY_arr.col(nc))
+            .sum();
+  }
+  for (size_t nc = 0; nc < N * C; ++nc) {
+    int c = nc % C;
+    dX_arr.col(nc) += scaleInvVarNHW(c) *
```
Comment (SherlockNoMad) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:92]: do we need += here?

Comment (SherlockNoMad) [reply orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:92]: dX_arr.setZero() is also probably not needed?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc (hunk 2 above)
Comment (pranav-prakash) [reply orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc]: Good catch, fixed.

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc (hunk 4 above)
Comment (pranav-prakash) [reply orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc]: Yes I think you're right – fixed.

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc (hunk 3 above)
Comment (pranav-prakash) [reply orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc]: Done, didn't realize that function existed!

**Review by SherlockNoMad — DISMISSED** [review]

**Comment by pranav-prakash** [comment]
@SherlockNoMad just to double-check to make sure I'm not missing something, you mentioned that
>Output4's name should actually be save_inv_variance.
But since we do in fact output "1/std_dev" isn't "saved_inv_std" correct? (where the "inverse" here denotes the multiplicative inverse i.e. reciprocal).

**Comment by SherlockNoMad** [comment]
For cpu kernel, the 4th forward output is calculating inv_std. 
For cuda kernel, the 4th forward output is inv_var, coz cudnn BatchNormForward is outputing inv_var. 

I think we will relax the BatchNormInternal's schema a bit to say, the 4th input can be either inv_std or inv_var. But forward kernel and backward kernel need to come in pair.

**Comment by mindest** [comment]
@pranav-prakash Hi Pranav, I found that when calculating the numerical jacobian, the Cpu/Cuda EPs are used sequentially and the Cpu results are covered by Cuda (see [this line](https://github.com/microsoft/onnxruntime/blob/ebee3809110f8c32ce5f77886fa1f4b4edf7806f/onnxruntime/test/providers/provider_test_utils.cc#L976)), but Cuda EP gives wrong answers in larger dim case because it seems that larger dim sizes (>5) are currently not supported (see "[NormalizeDims](https://github.com/microsoft/onnxruntime/blob/ebee3809110f8c32ce5f77886fa1f4b4edf7806f/onnxruntime/core/providers/cpu/nn/batch_norm_helper.h#L107)"). I tried locally and the test case can pass after some modifications of the NormalizeDims function.

As this seems to be related to Cuda implementation of BN, I think we could avoid such case (dim size > 5) for now, and resolve it later? @SherlockNoMad 

> I can't replicate the issue on my end using that same random seed though (`export ORT_TEST_RANDOM_SEED_VALUE=1914042201`) .

A simpler test case seems to repro the issue:

```cpp
    int channel_dim = 1;
    TensorShape in_out_shape({1, channel_dim, 1, 1, 1, 2});
    // ...
    std::vector<std::vector<float>> x_datas = {{-1.0f, 1.0f}, {1.0f}, {-1.0f}, {0.0f}, {1.0f}};
    gradient_checker.ComputeGradientError(op_def, {x_info, scale_info, bias_info, mean_info, var_info},
                                          {y_info, running_mean_info, running_var_info, saved_mean_info, saved_var_info}, &max_error,
                                          x_datas,
                                          {MakeAttribute("epsilon", epsilon), MakeAttribute("momentum", momentum)},
                                          /* check_not_have_gradient =*/ true,
                                          /* check_not_have_shape_inferencing = */ false,
                                          &execution_providers);
```

**Comment by pranav-prakash** [comment]
@mindest Thank you for investigating! Aren't we already excluding the CUDA EP by doing

```
std::vector<std::unique_ptr<IExecutionProvider>> execution_providers;
execution_providers.push_back(DefaultCpuExecutionProvider());
```
though? The CUDA EP should never get registered in that case.

**Comment by mindest** [comment]
Yes, but when [calculating the numeric jacobian](https://github.com/microsoft/onnxruntime/blob/f97764432430cef27dfd6782338604ed1f0a0a14/orttraining/orttraining/test/gradient/gradient_checker.cc#L335), the `execution_providers` is not passed and the [op tester](https://github.com/microsoft/onnxruntime/blob/f97764432430cef27dfd6782338604ed1f0a0a14/orttraining/orttraining/test/gradient/gradient_checker.cc#L101) still [runs over every EP if available](https://github.com/microsoft/onnxruntime/blob/f97764432430cef27dfd6782338604ed1f0a0a14/onnxruntime/test/providers/provider_test_utils.cc#L887). As for BatchNorm case, both Cpu/Cuda EPs are available for op `BatchNomalization`.

@SherlockNoMad Should we also pass `execution_providers` to `ComputeNumericJacobianTranspose()`
https://github.com/microsoft/onnxruntime/blob/f97764432430cef27dfd6782338604ed1f0a0a14/orttraining/orttraining/test/gradient/gradient_checker.cc#L335 to limit its available EPs, instead of the default setting (all available EPs will be registered)?

**Comment by SherlockNoMad** [comment]
@mindest, yes Gradient op tester should follow OpTest's handling of execution_providers.

**Comment by pranav-prakash** [comment]
@SherlockNoMad I've asked @simonguozirui to take stewardship of this PR (for same reason described in https://github.com/microsoft/onnxruntime/pull/7536#issuecomment-846297368). Execution providers should now be excluded properly – could you please re-run CI?

**Comment by SherlockNoMad** [comment]
<div class="js-timeline-item js-timeline-progressive-focus-container" data-gid="MDEyOklzc3VlQ29tbWVudDgzNzg1ODgzOQ==" style="box-sizing: border-box; color: rgb(36, 41, 46); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="TimelineItem js-comment-container" data-gid="MDEyOklzc3VlQ29tbWVudDgzNzg1ODgzOQ==" data-url="/_render_node/MDEyOklzc3VlQ29tbWVudDgzNzg1ODgzOQ==/timeline/issue_comment?variables%5BdeferredCommentActions%5D=false" style="box-sizing: border-box; position: relative; display: flex; padding: 16px 0px; margin-left: 16px;"><div class=" timeline-comment-group js-minimizable-comment-group js-targetable-element TimelineItem-body my-0 " id="issuecomment-837858839" style="box-sizing: border-box; margin-top: 4px; margin-bottom: 0px !important; min-width: 0px; max-width: 100%; color: var(--color-timeline-text); flex: 1 1 auto;"><div class="ml-n3 timeline-comment unminimized-comment comment previewable-edit js-task-list-container editable-comment js-comment timeline-comment--caret reorderable-task-lists current-user" data-body-version="4dd1c498df49ca928ecef57d95e75f07975583ff238c31e2eab370804326228c" data-unfurl-hide-url="/content_reference_attachments/hide" style="box-sizing: border-box; margin-left: -16px !important; position: relative; color: var(--color-text-primary); background-color: var(--color-bg-primary); border-color: var(--color-box-border-info); border-top-style: ; border-top-width: ; border-right-style: ; border-right-width: ; border-bottom-style: ; border-bottom-width: ; border-left-style: ; border-left-width: ; border-image-source: ; border-image-slice: ; border-image-width: ; border-image-outset: ; border-image-repeat: ; border-radius: 6px;"><div class="edit-comment-hide" style="box-sizing: border-box;"><task-lists sortable="" style="box-sizing: border-box;">/azp run orttraining-linux-ci-pipeline,orttraining-mac-ci-pipeline,orttraining-linux-gpu-ci-pipeline,centos7_cpu,Linux CPU Minimal Build E2E CI Pipeline,Linux Nuphar CI Pipeline,MacOS NoContribops CI Pipeline, orttraining-distributed,orttraining-amd-gpu-ci-pipeline, orttraining-ortmodule, orttraining-ortmodule-distributed, orttraining-ortmodule-distributed</task-lists></div><form class="js-comment-update" id="issuecomment-837858839-edit-form" action="https://github.com/microsoft/onnxruntime/issue_comments/837858839" accept-charset="UTF-8" method="post" style="box-sizing: border-box;"></form></div></div></div></div><div class="js-timeline-item js-timeline-progressive-focus-container" data-gid="MDEyOklzc3VlQ29tbWVudDgzNzg1OTAxMw==" style="box-sizing: border-box; color: rgb(36, 41, 46); font-family: -apple-system, BlinkMacSystemFont, &quot;Segoe UI&quot;, Helvetica, Arial, sans-serif, &quot;Apple Color Emoji&quot;, &quot;Segoe UI Emoji&quot;; font-size: 14px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; white-space: normal; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><div class="TimelineItem js-comment-container" data-gid="MDEyOklzc3VlQ29tbWVudDgzNzg1OTAxMw==" data-url="/_render_node/MDEyOklzc3VlQ29tbWVudDgzNzg1OTAxMw==/timeline/issue_comment?variables%5BdeferredCommentActions%5D=false" style="box-sizing: border-box; position: relative; display: flex; padding: 16px 0px; margin-left: 16px;"><br class="Apple-interchange-newline">

</div></div>

File: orttraining/orttraining/test/gradient/gradient_checker.cc [hunk 5]
```diff
@@ -455,7 +457,7 @@ inline Status GradientChecker<X_T, Y_T, JAC_T>::ComputeGradientErrorInternal(
   InitJacobians(x_infos, y_infos, &jacobian_ns);
   // Compute numeric Jacobian.
   ORT_RETURN_IF_ERROR(ComputeNumericJacobianTranspose(
-      op_def, x_infos, y_infos, JAC_T{1e-3f}, x_datas, y_datas, &jacobian_ns, attributes));
+      op_def, x_infos, y_infos, JAC_T{1e-3f}, x_datas, y_datas, &jacobian_ns, attributes, execution_providers));
```
Comment (mindest) [inline orttraining/orttraining/test/gradient/gradient_checker.cc:460]: missing arg `add_shape` here, could be `true`.

**Comment by mindest** [comment]
There seems to be some issue on [this line](https://github.com/microsoft/onnxruntime/blob/182c06a14c4b238d042a4a1e30a2726db8fe5b73/onnxruntime/test/providers/provider_test_utils.cc#L861) after we provide `execution_providers`. Maybe we could change it referring to lines [884](https://github.com/microsoft/onnxruntime/blob/182c06a14c4b238d042a4a1e30a2726db8fe5b73/onnxruntime/test/providers/provider_test_utils.cc#L884)-911 and [line 953](https://github.com/microsoft/onnxruntime/blob/182c06a14c4b238d042a4a1e30a2726db8fe5b73/onnxruntime/test/providers/provider_test_utils.cc#L953) (or other better ideas).

**Comment by pranav-prakash** [comment]
@mindest
> There seems to be some issue on [this line]
Do you mean that we have to create a new EP rather than just registering the one that was passed to us? Why should this be an issue?

Maybe in that case for our purposes it would be easier to make use of the  EP exclusion mechanism and exclude CUDA using `excluded_provider_types ` instead.

**Comment by mindest** [comment]
@pranav-prakash 

> Do you mean that we have to create a new EP rather than just registering the one that was passed to us? Why should this be an issue?

I think it's because `execution_providers` will be used [later](https://github.com/microsoft/onnxruntime/blob/182c06a14c4b238d042a4a1e30a2726db8fe5b73/orttraining/orttraining/test/gradient/gradient_checker.cc#L376) and `std::move(entry)`  would affect its contents.

> Maybe in that case for our purposes it would be easier to make use of the EP exclusion mechanism and exclude CUDA using excluded_provider_types instead.

I agree that this is an alternative.

**Comment by pranav-prakash** [comment]
@mindest
> would affect its contents

Ah I see. In fact someone else ran into the same problem on an old PR https://github.com/microsoft/onnxruntime/pull/6045#discussion_r543775689

The conclusion there was similar:

>the root of the problem is that `GradientOpTester::Run()` (for some unknown reason) has an unused execution_providers argument, and attempting to use it causes problems because the execution providers end up being consumed multiple times. This part of the code looks like it needs to be refactored, but I wouldn't want that to derail this PR.

So yes for this PR I think the best way forward is to make use of `excluded_provider_types` instead; I'll check with @simonguozirui to see if he's free to contribute, else try to get permission from my employer to amend this PR. 

Btw given that this issue has arisen twice now, perhaps a separate bug should be filed for this.

**Comment by mindest** [comment]
@pranav-prakash  Hi Pranav, since there is not much bandwidth for you to update the PR, is it okay that we take over? Thanks!

**Comment by pranav-prakash** [comment]
@mindest Yes that would be great, thank you

**Comment by mindest** [comment]
> In fact someone else ran into the same problem on an old PR [#6045 (comment)](https://github.com/microsoft/onnxruntime/pull/6045#discussion_r543775689)
> 
> The conclusion there was similar:
> 
> > the root of the problem is that `GradientOpTester::Run()` (for some unknown reason) has an unused execution_providers argument, and attempting to use it causes problems because the execution providers end up being consumed multiple times. This part of the code looks like it needs to be refactored, but I wouldn't want that to derail this PR.
> 
> So yes for this PR I think the best way forward is to make use of `excluded_provider_types` instead; I'll check with @simonguozirui to see if he's free to contribute, else try to get permission from my employer to amend this PR.
> 
> Btw given that this issue has arisen twice now, perhaps a separate bug should be filed for this.

Based on Pranav and Simon's contribution, some changes related to the above discussion are reverted in recent commits -- since cuda EP is now fixed, CI can run without throwing errors. But the above problem still exists and may need further refactor work, which is not BatchNorm specific.

**Comment by mindest** [comment]
azp run Linux CPU CI Pipeline,Linux CPU Minimal Build E2E CI Pipeline,Linux GPU CI Pipeline,Linux GPU TensorRT CI Pipeline,Linux OpenVINO CI Pipeline,Linux QNN CI Pipeline,MacOS CI Pipeline,ONNX Runtime Web CI Pipeline

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 6]
```diff
@@ -0,0 +1,102 @@
+/**
+* Copyright (c) 2016-present, Facebook, Inc.
```
Comment (mindest) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:2]: Should we remove the license here? @pengwa

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.h [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+namespace contrib {
+
+template <typename T>
+class BatchNormalizationGrad final : public OpKernel {
+ public:
+  explicit BatchNormalizationGrad(const OpKernelInfo& info) : OpKernel(info)  {}
+
+  Status Compute(OpKernelContext* context) const override;
+
+  private:
```
Comment (mindest) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.h:20]: (todo) nit: identation

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.h [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+#include "core/util/math.h"
+#include "core/mlas/inc/mlas.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+template <typename T>
+class BatchNormalizationGrad final : public OpKernel {
+ public:
+  explicit BatchNormalizationGrad(const OpKernelInfo& info) : OpKernel(info)  {}
```
Comment (mindest) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.h:16]: (todo) nit: space

File: orttraining/orttraining/test/training_ops/cpu/nn/batchnorm_internal_test.cc [hunk 9]
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/framework/tensor.h"
+#include "core/session/inference_session.h"
+#include "test/providers/provider_test_utils.h"
+
+#include "gtest/gtest.h"
+#include "gmock/gmock.h"
```
Comment (pengwa) [inline orttraining/orttraining/test/training_ops/cpu/nn/batchnorm_internal_test.cc:9]: are those includes are needed?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+#include "core/providers/cpu/nn/batch_norm_helper.h"
+
+namespace onnxruntime {
+namespace contrib {
+
+template <typename T>
+Status BatchNormalizationGrad<T>::Compute(OpKernelContext* ctx) const {
+  const Tensor* dY = ctx->Input<Tensor>(0);
+  const Tensor* X = ctx->Input<Tensor>(1);
+  const Tensor* Scale = ctx->Input<Tensor>(2);
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:33]: nit: Scale->scale?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+
+  const TensorShape X_shape = X->Shape();
+  const TensorShape channel_shape = saved_mean->Shape();
+
+  // no B here, but B has same size as Scale, so can validate inputs for gradient with this substitute
+  ORT_RETURN_IF_ERROR(BatchNormHelper::ValidateInputs(X, Scale, Scale, saved_mean, saved_inv_std));
+
+  const auto* dY_data = dY->template Data<T>();
+  const auto* X_data = X->template Data<T>();
+  const auto* Scale_data = Scale->template Data<T>();
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:45]: Scale_data looks weird, shall we change to scale_data?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  ConstEigenVectorArrayMap<T> mean_arr(saved_mean_data, scale_tensor_size);
+  ConstEigenVectorArrayMap<T> inv_std_arr(saved_inv_std_data, scale_tensor_size);
+
+  EigenVectorArrayMap<T> dBias_arr(dBias_data, scale_tensor_size);
+  EigenVectorArrayMap<T> dScale_arr(dScale_data, scale_tensor_size);
+
+  dBias_arr.setZero();
+  dScale_arr.setZero();
+
+  const auto scaled_inv_std = scale_arr * inv_std_arr / (N * sample_size);
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:71]: is it safe we just do a div and not check it is divisble or not?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+  return Status::OK();
+}
+
+ONNX_OPERATOR_KERNEL_EX(
+    BatchNormalizationGrad,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:98]: register type for T1 and T2? 

```
  ONNX_CONTRIB_OPERATOR_SCHEMA(BatchNormalizationGrad)
      .SetDomain(kMSDomain)
      .SinceVersion(1)
      .SetDoc("BatchNormalizationGrad")
      .Attr("epsilon",
            "epsilon value",
            AttributeProto::FLOAT)
      .Input(0, "dY", "Gradient output from previous node", "T")
      .Input(1, "X", "Input", "T")
      .Input(2, "scale", "Scale tensor", "T1")
      .Input(3, "mean", "Mean of X", "T2")
      .Input(4, "variance", "Variance of X", "T2")
      .Output(0, "X_grad", "Gradient of the input", "T")
      .Output(1, "scale_grad", "Gradient of the scale", "T1")
      .Output(2, "bias_grad", "Gradient of the bias", "T1")
      .TypeConstraint(
          "T",
          {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
          "Constrain input and output types to float tensors.")
      .TypeConstraint(
          "T1",
          {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
          "Constrain scale and bias types to float tensors.")
      .TypeConstraint(
          "T2",
          {"tensor(float16)", "tensor(float)", "tensor(double)", "tensor(bfloat16)"},
          "Constrain mean and variance types to float tensors.");
```

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_internal.cc [hunk 14]
```diff
@@ -0,0 +1,33 @@
+/**
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_internal.cc:1]: I think we can remove this.

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_internal.cc [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+
+namespace onnxruntime {
+namespace contrib {
+
+ONNX_OPERATOR_KERNEL_EX(
+    BatchNormInternal,
+    kMSDomain,
+    1,
+    kCpuExecutionProvider,
+    KernelDefBuilder().Alias(3,1).Alias(4,2).TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_internal.cc:29]: consider to register T1 and T2 also.

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_internal.cc [hunk 16]
```diff
@@ -0,0 +1,33 @@
...
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+/* Modifications Copyright (c) Microsoft. */
+
+#include "core/providers/cpu/nn/batch_norm.h"
+#include "core/providers/cpu/nn/batch_norm_helper.h"
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_internal.cc:19]: the include is useless?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+* limitations under the License.
+*/
+/* Modifications Copyright (c) Microsoft. */
+
+#include "batch_norm_grad.h"
+#include "core/util/math.h"
+#include "core/util/math_cpuonly.h"
+#include "core/providers/common.h"
+#include "core/framework/op_kernel_context_internal.h"
+#include "core/common/safeint.h"
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:23]: consider to remove those useless include.

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc [hunk 18]
```diff
@@ -0,0 +1,102 @@
...
+  const Tensor* dY = ctx->Input<Tensor>(0);
+  const Tensor* X = ctx->Input<Tensor>(1);
+  const Tensor* Scale = ctx->Input<Tensor>(2);
+  const Tensor* saved_mean = ctx->Input<Tensor>(3);
+  const Tensor* saved_inv_std = ctx->Input<Tensor>(4);
+
+  const TensorShape X_shape = X->Shape();
+  const TensorShape channel_shape = saved_mean->Shape();
+
+  // no B here, but B has same size as Scale, so can validate inputs for gradient with this substitute
```
Comment (pengwa) [inline orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:40]: this looks to be a workaround, maybe we can just check pointer B is null or not in the ValidateInputs?

File: orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc (hunk 12 above)
Comment (mindest) [reply orttraining/orttraining/training_ops/cpu/nn/batch_norm_grad.cc:71]: It should be ok because `scale_arr` and `inv_std_arr` are floats.

**Review by baijumeswani — APPROVED** [review]

**Review by pengwa — APPROVED** [review]
:shipit:

**Comment by mindest** [comment]
Thanks! @pengwa @baijumeswani

**Comment by mindest** [comment]
@pranav-prakash thanks for your contribution!

**Pull request merged**